### ISSUE TITLE
There should be a region endpoint that allows all regions to be retrieved without an SDK update.

### ISSUE THREAD

**Issue by rdifalco** [aws/aws-sdk-java#948]
We have code that iterates all regions and (for example) calls describe instance on each for a given credential. The SDK should provide region enums as a convenience, not a necessity.  All clients should work with new regions that show up. Partition files should be pulled from an URL instead of json backed into the SDK build. Then `RegionUtils.getRegions()` would return all current regions (or at least as current as the last time that process started. If the regions URL is not available then it could fall back onto the backed in partition json files.

**Comment by shorea** [comment]
That's an interesting suggestion. What kind of application are you building (if you don't mind sharing that is)?

**Comment by rdifalco** [comment]
We monitor cost, make RI recommendations, track utilization, etc. for our customers. But even for our own infrastructure it would be nice to be able to discover instances or RIs across all current regions.

**Comment by fernomac** [comment]
That's how region metadata worked originally. Caused quite a kerfuffle when new regions started showing up that required SigV4, and people using old SDKs that didn't know suddenly started trying to connect to them and throwing authentication exceptions. Tread carefully.

**Comment by rdifalco** [comment]
Well, supported signature versions are in the metadata.

**Comment by fernomac** [comment]
They are now. Guess when that got added. :)

I'm all for a proper, versioned service API exposing region/endpoint metadata, and support in the SDK for consuming it. But be careful - writing code today that will be forward-compatible with all the crazy stuff that AWS might dream up for future region builds is hard.

Cutting out the middleman, what'd be *really* nice is a 'DoItInAllRegionsIDontCareIfItsSlowAndMightFail' parameter on describe calls, and having the service manage the details for me...

**Comment by rdifalco** [comment]
If AWS is too squeamish about adding currently available regions in `RegionUtils#getRegions` I _really_ wouldn't want a `DoItInAllRegionsIDontCareIfItsSlowAndMightFail`.  I much prefer a method that returns a list of regions like `getRegions` but with a name like `getAllSupportedRegions` with a code comment that this may not work with all clients for regions and/or clients that require special Java AWS SDK support.

**Comment by kiiadi** [comment]
Sorry about the radio silence on this one. Just to understand a bit more about the ask, for a given service you want to know what regions that service is available in without having to grab the latest SDK - is that right?

The possibility of adding this as a new service is an interesting one.

I know it's not the same, but as a temporary *hack* could you scrape this page : https://aws.amazon.com/about-aws/global-infrastructure/regional-product-services/ ?

**Comment by kiiadi** [comment]
Another option might be to look at the endpoints json in our github repo: https://raw.githubusercontent.com/aws/aws-sdk-java/master/aws-java-sdk-core/src/main/resources/com/amazonaws/partitions/endpoints.json

**Comment by rdifalco** [comment]
Yes @kiiadi, those are great suggestions for workarounds, thanks!

**Comment by rdifalco** [comment]
@kiiadi, I think I could also get it from the pricing sheet end point, though a little heavy handed (lots of data to just get the regions). The only problem with the endpoints.json in the aws-sdk-java repo is that it could show up in the price sheet before it is addressed in the SDK. But not a huge issue as those are usually turned around pretty quickly.

**Comment by kiiadi** [comment]
@rdifalco you're right there might be a slight delay between when a service gets released in a given region and when it shows up in the endpoints.json file, but usually no more than a few days.

**Comment by kiiadi** [comment]
Obviously you're going to want to be defensive about linking to a URL that's hosted in a github repository - it may break for a variety of reasons including:
1. github throttling requests
2. it being moved inside the repo (we don't have any current plans to move the [endpoint.json](https://raw.githubusercontent.com/aws/aws-sdk-java/master/aws-java-sdk-core/src/main/resources/com/amazonaws/partitions/endpoints.json) file but it is something of an internal concern)

If you do choose to go this route, I recommend implementing some disk cache in the event that you can't load the latest endpoint.json from github.

You can potentially re-use the same model object that the SDK uses to parse the file that ships with the SDK via Jackson - see [Partitions.java](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-core/src/main/java/com/amazonaws/partitions/model/Partitions.java). A "dumb" implementation might look like this:

```java
public static Map<String, Set<String>> getRegions() throws IOException {
  InputStream is = URI.create(endpointUrl).toURL().openStream();
  Partitions partitions = objectMapper.readValue(is, Partitions.class);

  Partition awsPartition = partitions.getPartitions().stream()
      .filter(p -> "aws".equals(p.getPartition()))
      .findFirst()
      .orElseThrow(() -> new RuntimeException("Could not find AWS partition"));

  return awsPartition.getServices().entrySet().stream()
      .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().getEndpoints().keySet()));
}
```

**Comment by rdifalco** [comment]
@kiiadi  the problem with this solution is that there was a 15 day lag between when us-east-2 was introduced and when it showed up in this file.

**Comment by debora-ito** [comment]
This is still an interesting suggestion, but a modeled service API is a better place for this, then all the language SDKs can get a generated client from it.

I'll move this to the cross-SDK repo to see if we can get some traction.
